### PR TITLE
Add nested table support

### PR DESCRIPTION
**Pull request by lahirusamith**
## Purpose
$title

Fixes ballerina-platform/ballerina-standard-library#1665

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

**Comment by daneshk** [comment]
@lahirusamith we need negative test cases to validate error messages

File: ballerina/types.bal [hunk 1]
```diff
@@ -68,6 +68,15 @@ public type Varray record {|
...
+# + name - Name of the varray
+# + elements - Elements of the Varray
+public type NestedTableType record {|
+    string name;
+    ArrayValueType? elements;
+|};
+
 # Represents OBJECT TYPE Oracle DB field.
 #
 # + value - Value of parameter passed into the SQL statement
```
Comment (daneshk) [inline ballerina/types.bal:82]: Do we use these object types anywhere

**Comment by lahirusamith** [comment]
> @lahirusamith we need negative test cases to validate error messages

@daneshk few negative test cases added in https://github.com/ballerina-platform/module-ballerinax-oracledb/pull/32/commits/f266089218ccdad56b920b0a2764333204ca3249.

File: ballerina/types.bal (hunk 1 above)
Comment (lahirusamith) [reply ballerina/types.bal:82]: Object types are only used to pass the parameters to the prepared statement. when retrieving data we are not using object types.

File: native/src/main/java/io/ballerina/stdlib/oracledb/utils/ConverterUtils.java [hunk 2]
```diff
@@ -229,7 +229,7 @@ public static BArray convertByteArrayToBallerinaType(DefaultResultParameterProce
             }
             return byteDataArray;
         } else {
-            throw new UnsupportedTypeError(String.format("Cannot cast varray to type: %s", type));
+            throw new UnsupportedTypeError(String.format("Cannot cast array to type: %s", type));
```
Comment (niveathika) [inline native/src/main/java/io/ballerina/stdlib/oracledb/utils/ConverterUtils.java:232]: Why is the default parameter processor passed to this method? This is not standard practise

**Review by niveathika — CHANGES_REQUESTED** [review]

File: native/src/main/java/io/ballerina/stdlib/oracledb/utils/ConverterUtils.java (hunk 2 above)
Comment (lahirusamith) [reply native/src/main/java/io/ballerina/stdlib/oracledb/utils/ConverterUtils.java]: Moved few methods in https://github.com/ballerina-platform/module-ballerinax-oracledb/pull/32/commits/38efc04030ca7722f72f8743b9f51d1e7219421c

**Review by kaneeldias — APPROVED** [review]

File: ballerina/tests/Config.toml [hunk 3]
```diff
@@ -16,6 +16,6 @@
 # under the License.
 #
 [ballerina.sql]
-maxOpenConnections=10
+maxOpenConnections=15
```
Comment (daneshk) [inline ballerina/tests/Config.toml:19]: Why this is increased?

Comment (lahirusamith) [reply ballerina/tests/Config.toml]: Revert back to 10 by ordering few test cases in different test groups.

File: ballerina/tests/11-custom-object-type-test.bal [hunk 4]
```diff
@@ -371,13 +370,105 @@ isolated function selectObjectTypeWithInvalidTypedRecord() returns error? {
...
+        int[] gradesAtt = [67, 45, 78, 86];
+        NestedTableValue grades = new({name: "NestedGradeTable", elements: gradesAtt});
+        int total = 4;
+        int pk = 2;
+        string teacher = "Kate Anderson";
+        sql:ParameterizedQuery insertQuery = `INSERT INTO NestedClassTable(pk, col_teacher, col_students, col_grades, col_total)
+                VALUES (${pk}, ${teacher}, ${students}, ${grades}, ${total})`;
+        sql:ExecutionResult result = check executeQuery(insertQuery);
+        test:assertExactEquals(result.affectedRowCount, 1, "Affected row count is different.");
+        var insertId = result.lastInsertId;
```
Comment (niveathika) [inline ballerina/tests/11-custom-object-type-test.bal:399]: Format

Comment (lahirusamith) [reply ballerina/tests/11-custom-object-type-test.bal]: done in https://github.com/ballerina-platform/module-ballerinax-oracledb/pull/32/commits/4eeb31cc93af8617db71bac6ca0db26dc62e0863

File: ballerina/tests/11-custom-object-type-test.bal [hunk 5]
```diff
@@ -371,13 +370,105 @@ isolated function selectObjectTypeWithInvalidTypedRecord() returns error? {
...
+}
+isolated function insertNestedTableNull() returns error? {
+    int pk = 3;
+    NestedTableValue students = new();
+    NestedTableValue grades = new();
+    sql:ParameterizedQuery insertQuery = `INSERT INTO NestedClassTable(pk, col_students, col_grades)
+            VALUES (${pk}, ${students}, ${grades})`;
+    sql:ExecutionResult|sql:Error result = executeQuery(insertQuery);
+    if result is sql:ApplicationError {
+       test:assertTrue(result.message().includes("Invalid parameter: null is passed as value for SQL type: varray"));
```
Comment (daneshk) [inline ballerina/tests/11-custom-object-type-test.bal:445]: In the error message, the SQL type should be `nested table` right?

**Review by niveathika — APPROVED** [review]

**Review by daneshk — APPROVED** [review]
LGTM. Fix the conflicts

**Pull request merged**